## 1 Reading CSV Files with Encodings

So far, we've learned how to select, assign, and analyze data with pandas using pre-cleaned data. In reality, data is rarely in the format you need it to be to perform analysis. Data scientists commonly [spend over half their time cleaning data](https://www.forbes.com/sites/gilpress/2016/03/23/data-preparation-most-time-consuming-least-enjoyable-data-science-task-survey-says/#20705a446f63), so knowing how to clean 'messy' data is an extremely important skill.

In this lesson, you'll learn how to :

- Clean columns names.
- Extract and convert numeric values from string values.
- Extract string data.
- Work with missing values.

We'll be working with **laptops.csv**, a CSV file containing information on about 1,300 laptop computers. The first five rows of the CSV file is shown below:

| _ | Manufacturer | Model Name  | Category  | Screen Size | Screen                             | CPU                        | RAM  | Storage             | GPU                          | Operating System | Operating System Version | Weight | Price (Euros) |
|---|--------------|-------------|-----------|-------------|------------------------------------|----------------------------|------|---------------------|------------------------------|------------------|--------------------------|--------|---------------|
| 0 | Apple        | MacBook Pro | Ultrabook | 13.3"       | IPS Panel Retina Display 2560x1600 | Intel Core i5 2.3GHz       | 8GB  | 128GB SSD           | Intel Iris Plus Graphics 640 | macOS            | NaN                      | 1.37kg | 1339,69       |
| 1 | Apple        | Macbook Air | Ultrabook | 13.3"       | 1440x900                           | Intel Core i5 1.8GHz       | 8GB  | 128GB Flash Storage | Intel HD Graphics 6000       | macOS            | NaN                      | 1.34kg | 898,94        |
| 2 | HP           | 250 G6      | Notebook  | 15.6"       | Full HD 1920x1080                  | Intel Core i5 7200U 2.5GHz | 8GB  | 256GB SSD           | Intel HD Graphics 620        | No OS            | NaN                      | 1.86kg | 575,00        |
| 3 | Apple        | MacBook Pro | Ultrabook | 15.4"       | IPS Panel Retina Display 2880x1800 | Intel Core i7 2.7GHz       | 16GB | 512GB SSD           | AMD Radeon Pro 455           | macOS            | NaN                      | 1.83kg | 2537,45       |
| 4 | Apple        | MacBook Pro | Ultrabook | 13.3"       | IPS Panel Retina Display 2560x1600 | Intel Core i5 3.1GHz       | 8GB  | 256GB SSD           | Intel Iris Plus Graphics 650 | macOS            | NaN                      | 1.37kg | 1803,60       |

We can start by reading the data into pandas. 


In [0]:
# Did you get an error?
import pandas as pd
url = "https://raw.githubusercontent.com/nymarya/data-science-one/master/Lesson%2305/laptops.csv"
laptops = pd.read_csv(url)
laptops.head()


UnicodeDecodeError: ignored

We get an error! (The actual error is much longer than this, we've truncated it for readability). To understand this error and how to fix it, we need to learn about **encodings**.

**A Short History of Encodings**

Computers, at their lowest levels, can only understand binary - 0 and 1. Encodings are systems for representing characters in binary. From the early days of computers, the standard for representing text was called ASCII. The ASCII standard specified a set of 128 standard characters - these were letters, numbers and punctuation marks that were used for the English language. For instance, the letter __a__ in ASCII is represented as **01100001** in binary.

ASCII was very useful for the basic letters and numbers in the english language, but as the popularity of computers spread worldwide, different computer manufacturers created new encodings to encode different characters, like the greek character $\alpha$ or the Japanse character も. Because there wasn't one standard for encodings there were problems when you tried to read files saved using one character set with a computer that used a different character set.


<img width="600" src="https://drive.google.com/uc?export=view&id=1NzDcepAjvJESItrSbykVB3L9SFdLCLaS">

As a result, there ended up being dozens of different encodings being used throughout the early days of the web. Ideally, the person sending you a file would tell you what encoding it is in, however in reality that rarely happened (and still doesn't). Trying to work out what encoding a file was in was relatively difficult, and you often had to guess.

These days, we often still have to guess, however in the last 15 years the number of encodings being commonly used has decreased. This means that the vast majority of files are encoded in one of 2-3 formats.


<img width="600" src="https://drive.google.com/uc?export=view&id=1mPe_Flzn-VfBCZht69_k2rpopYpHqHYY">


The diagram above shows the usage of different encodings on the web. As you can see, **UTF-8** has grown rapidly to be the predominant encoding being used. Because of this trend, the best thing to do if your file has an unknown encoding is to try the most common encodings. The most common encodings are, in order:

- UTF-8 (the default for Python)
- Latin-1 (also known as ISO-8895-1)
- Windows-1251

To specify a encoding when reading a CSV file with pandas, simply use the **encoding** argument within the **pandas.read_csv()** function, specifying the encoding as a string:

```python
df = pd.read_csv("filename.csv", encoding="UTF-8")
```


Because **UTF-8** is the default, you don't need to specify it the file you're reading is encoded with **UTF-8** (you'll notice the error message mentions **UTF-8**)


**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>


- Use the **pandas.read_csv()** function to read the **laptops.csv** file into a dataframe **laptops**.
  - Specify the encoding using the string **"Latin-1"**.
  - If that doesn't work, try using the string **"Windows-1251"**.
- Use the **DataFrame.info()** method to display information about the laptops dataframe.

In [0]:
# put your code here 
laptops = pd.read_csv(url, encoding='latin-1')

print(laptops.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 13 columns):
Manufacturer                1303 non-null object
Model Name                  1303 non-null object
Category                    1303 non-null object
Screen Size                 1303 non-null object
Screen                      1303 non-null object
CPU                         1303 non-null object
RAM                         1303 non-null object
 Storage                    1303 non-null object
GPU                         1303 non-null object
Operating System            1303 non-null object
Operating System Version    1133 non-null object
Weight                      1303 non-null object
Price (Euros)               1303 non-null object
dtypes: object(13)
memory usage: 132.4+ KB
None


## 2 Cleaning Column Names

In [0]:
laptops.columns.tolist()

['Manufacturer',
 'Model Name',
 'Category',
 'Screen Size',
 'Screen',
 'CPU',
 'RAM',
 ' Storage',
 'GPU',
 'Operating System',
 'Operating System Version',
 'Weight',
 'Price (Euros)']

As well as being able to use the attribute to view the column labels, we can also assign to the attribute:


In [0]:
laptops_test = laptops.copy()
laptops_test.columns = ['A', 'B', 'C', 'D', 'E',
                        'G', 'F', 'H', 'I', 'J',
                        'K', 'L', 'M']
laptops_test.columns.tolist()

['A', 'B', 'C', 'D', 'E', 'G', 'F', 'H', 'I', 'J', 'K', 'L', 'M']

We can create a function that uses Python [string methods](https://docs.python.org/3/library/stdtypes.html#string-methods) to clean our column labels, and then use list comprehension to apply that function to each label. Let's look at an example:




In [0]:
def clean_col(col):
    col = col.strip()
    col = col.replace("(","")
    col = col.replace(")","")
    col = col.lower()
    return col

laptops.columns = [clean_col(c) for c in laptops.columns]
laptops.columns.tolist()

['manufacturer',
 'model name',
 'category',
 'screen size',
 'screen',
 'cpu',
 'ram',
 'storage',
 'gpu',
 'operating system',
 'operating system version',
 'weight',
 'price euros']

Let's use this technique to clean the column labels in our dataframe, adding a few extra cleaning 'chores' along the way

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

1. Define a function, which accepts a string argument, and:
  - Removes any whitespace from the start and end of the string.
  - Replaces the substring Operating System with the abbreviation os.
  - Replaces all spaces with underscores.
  - Removes parentheses from the string.
  - Makes the entire string lowercase.
  - Returns the modified string.
2. Use list comprehension to apply the function to each item in the **DataFrame.column** attribute for the **laptops** dataframe, assigning the result back to the the **DataFrame.columns** attribute.

In [0]:
# put your code here
def convert(string):
  string = string.strip()
  string = string.replace("perating system", "S")
  string = string.replace(" ", "_")
  string = string.replace(")", "")
  string = string.replace("(", "")
  string = string.lower()
  return string


laptops.columns = [convert(c) for c in laptops.columns]
laptops.columns.tolist()

['manufacturer',
 'model_name',
 'category',
 'screen_size',
 'screen',
 'cpu',
 'ram',
 'storage',
 'gpu',
 'os',
 'os_version',
 'weight',
 'price_euros']

## 3 Converting String Columns to Numeric

In [0]:
laptops.iloc[:5,2:5]

category screen_size                              screen
0  Ultrabook       13.3"  IPS Panel Retina Display 2560x1600
1  Ultrabook       13.3"                            1440x900
2   Notebook       15.6"                   Full HD 1920x1080
3  Ultrabook       15.4"  IPS Panel Retina Display 2880x1800
4  Ultrabook       13.3"  IPS Panel Retina Display 2560x1600

Of these three columns, we have three different types of text data:

- **category** - This is purely text data, there are no numeric values.
- **screen_size** - This is numeric data that is being stored as text data because of the " character.
- **screen** - This is a combination of pure text data with numeric data.


Thinking about the **screen_size** column specifically, while the values are stored as text data, we can't sort them numerically. For instance, if we wanted to select laptops with screens 15" or larger we're unable to do so. To be able to answer these sorts of questions we need to convert the string values to numeric values.

Whenever we're converting text to numeric data, we can follow this data cleaning workflow:

<img width="500" src="https://drive.google.com/uc?export=view&id=12C8ysc6Sd5BcLmpKwIJzlMtVB6_SPQ1D">


Let's walk through the workflow while we convert the **screen_size** column to numeric. The first stage is to explore the data. One of the best ways to do this is to use the **Series.unique()** method to view all of the unique values in the colum.

In [0]:
print(laptops["screen_size"].dtype)
laptops["screen_size"].unique()

Our next stage is **to identify patterns and special cases**. We can see that all values in this column follow the same pattern - series of digit and period characters, followed by a quote character. There are no special cases - every value matches the same pattern. We can also observe that we will need to convert the column to a **float dtype**, as the **int dtype** won't be able to store the decimal values.

The next stage is **to remove the non-digit characters**. The pandas library contains dozens of vectorized string methods, most of which are available using the **Series.str accessor**. In this case, we can use the **Series.str.replace()** method, which is a vectorized version of the **Python str.replace()** method we used in the previous screen to remove all the quote characters:

In [0]:
laptops["screen_size"] = laptops["screen_size"].str.replace('"','')
laptops["screen_size"].unique()

Now we've **removed the non-digit characters**, we can **convert (or cast)** the column to a numeric dtype. To do this, we use the **Series.astype()** method. We can use either **int** or **float** as the parameter for the method to convert the column to the respective type:




In [0]:
laptops["screen_size"] = laptops["screen_size"].astype(float)
print(laptops["screen_size"].dtype)
laptops["screen_size"].unique()

Our column is now the **float64** dtype, and you see that there are no longer quotes around each value denoting them as strings.

Our final step is **to rename** the column. This is an optional step, and can be useful if the non-digit values contained information that helps us understand the data. In out case the quote characters actually denoted that the screen size was in inches. We can use the **DataFrame.rename()** method to rename specific axis labels using a dictionary with the keys as the old label name, and the values as the new label name. We'll also need to specify **axis=1** parameter so pandas knows that we want to rename labels in the column axis, and we'll also use **inplace=True** instead of assignment (although assigning back to the DataFrame would give us an identical result):

In [0]:
laptops.rename({"screen_size": "screen_size_inches"}, axis=1, inplace=True)
laptops.dtypes

We can see that we now have one column converted to a numeric type and renamed appropriately. Our exercise for this lesson will be to follow the process for the **ram** column. We'll do the first two steps together, starting with exploring the data:






In [0]:
laptops["ram"].unique()

We can easily identify a clear pattern to the data - all values are integers, and include the character **GB** at the end of the string. There aren't any special cases that are exceptions to the pattern.

Let's finish the process of converting the **ram** column to a numeric type.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

1. Use the **Series.replace()** method to remove the substring **GB** from the ram column.
2. Use the **Series.astype()** method to change the **ram** column to an **integer** dtype.
3. Because the **GB** characters contained useful information about the units (gigabytes) of the laptop's ram, use the **DataFrame.rename()** method to rename the column from **ram** to **ram_gb.**
4. Use the **DataFrame.dtypes** attribute to get a list of the column names and types from the **laptops** dataframe, and assign the result to **dtypes**.

In [0]:
# put your code here
laptops["ram"] = laptops["ram"].str.replace("GB", "").astype(int)
laptops.rename({"ram": "ram_gb"}, axis=1, inplace=True)

dtypes = laptops.dtypes
print(dtypes)

manufacturer    object
model_name      object
category        object
screen_size     object
screen          object
cpu             object
ram_gb           int64
storage         object
gpu             object
os              object
os_version      object
weight          object
price_euros     object
dtype: object


## 4 Practicing Converting String Columns to Numeric



This workflow is extremely common, so we're going to practice it with the **weight** and **price_euros** columns. Here's a reminder of the workflow

<img width="500" src="https://drive.google.com/uc?export=view&id=12C8ysc6Sd5BcLmpKwIJzlMtVB6_SPQ1D">

Let's start with the **weight** column:



In [0]:
laptops.weight.unique()

While it appears that the **weight** column may just need the **kg** characters removed from the end of each string, there are a lot of unique values for the weight column, so it's hard to visually confirm if there are any exceptions to the pattern.

If we can't see any exceptions, it's OK to move forward onto the next step, as if we miss any, the error we get will tell us the value so we can fix it. Let's see if we missed anything– we're going to use method chaining to attempt both removing the **kg** characters and casting to the **float**, but we'll use a new formatting trick for method chaining. By putting our code inside parentheses, we can method chain over multiple lines, which makes our code easier to read.





In [0]:

laptops["weight"] = (laptops["weight"]
                     .str.replace("kg","")
                     .astype(float)
                    )

We've hit an error– **a ValueError** to be precise (we've truncated the error output to make it easier to read, but you will always find the value in question at the very bottom of the error). The error tells us which value it couldn't convert to a float **'4s'**. Keep in mind that this is the value after the **kg** has been replaced because of our method chaining, so the value substring **'4s'** may not actually exist in the raw data. We can use the pandas **Series.str.contains()** method, which returns a boolean series based on whether a substring is found to look at the raw value:

In [0]:
laptops.loc[laptops["weight"].str.contains('s'), "weight"]

KeyError: ignored

We have identified the special case - we need to remove **kgs** before **kg**, or alternatively we could remove **kg** and then **s**. There's the possibility that there might be more than one exception, because the error will stop the rest of the values from being processed. We'll leave that final step up to you in the exercise at the end of this screen.

Next, let's look at the **price_euros** column. This column has almost 800 unique values, so we're going to look at the first and last 5 as a sample:




In [0]:
print(laptops["price_euros"].unique()[:5])
print(laptops["price_euros"].unique()[-5:])

['1339,69' '898,94' '575,00' '2537,45' '1803,60']
['549,99' '805,99' '720,32' '638,00' '764,00']


You may not have seen this style of writing decimals with a comma instead of a period, which is used in parts of South America, Africa and Europe (you might like to read more about [decimal commas](https://en.wikipedia.org/wiki/Decimal_separator#Hindu%E2%80%93Arabic_numeral_system)). In this case, simply removing the comma will give us incorrect values– we'll need to replace with with a period.

Now that we've done some exploration together, let's finish off the cleaning of both of these columns.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>


1. Clean the **weight** column by:
  - Removing all non-digit characters.
  - Casting the column to the appropriate numeric type
  - Renaming the column to **weight_kg**.
2. Clean the **price_euros** column by:
  - Replacing the decimal commas with decimal points.
  - Casting the column to the appropriate numeric type.
3. Use the **Series.describe()** method to generate some descriptive statistics for each column:
  - Assign the results for the **weight_kg** column to **weight_describe**.
  - Assign the results for the **price_euros** column to **price_describe**.

In [0]:
# put your code here

laptops['weight'] = laptops['weight'].str.replace("kg", "")
laptops['weight'] = laptops['weight'].str.replace("s", "").astype(float)
laptops.rename({"weight": "weight_kg"}, axis=1, inplace=True)

weight_describe = laptops['weight_kg'].describe()

print(weight_describe)

KeyError: ignored

In [0]:
laptops['price_euros'] = laptops['price_euros'].str.replace(",",".").astype(float)

price_describe = laptops['price_euros'].describe()

print(price_describe)

count    1303.000000
mean     1123.686992
std       699.009043
min       174.000000
25%       599.000000
50%       977.000000
75%      1487.880000
max      6099.000000
Name: price_euros, dtype: float64



## 5 Extracting Values from the Start of Strings



From the previous screen, we can see that the average (mean) laptop weights around 2kg and costs just over 1,100 euros.

Sometimes, it can be useful to extract non-numeric values from within strings. Let's look at the first 10 values from the **gpu** (graphics processing unit) column:


In [0]:
laptops["gpu"].head(10)

0    Intel Iris Plus Graphics 640
1          Intel HD Graphics 6000
2           Intel HD Graphics 620
3              AMD Radeon Pro 455
4    Intel Iris Plus Graphics 650
5                   AMD Radeon R5
6         Intel Iris Pro Graphics
7          Intel HD Graphics 6000
8            Nvidia GeForce MX150
9          Intel UHD Graphics 620
Name: gpu, dtype: object

The information in this column seems to be a manufacturer followed by a model name/number. The manufacturers in the subset are just the first word of each value: Intel, AMD, Nvidia. Extracting the manufacturer by itself would be useful, as we can analyze who are the most common GPU manufacturers.

Because each manufacturer is followed by a whitespace character, we can use the **Series.str.split()** method to extract this data. Let's look how we use this method to extract the data.

First, we'll use **Series.head()** to look at just the first five items:

<img width="500" src="https://drive.google.com/uc?export=view&id=1cjQOzRy2pFfEntvcoQY8zzZwi-OcfWH2">

Net's we'll start by using **Series.str.split()** with the default parameters:

<img width="500" src="https://drive.google.com/uc?export=view&id=1j_p4vBq4DH6MqmqpLrDCc6aWvxEJztkg">

The method has split each string on the whitespace, and the result is individual Python lists stored within a series. The **Series.str.split()** method accepts an argument **n**, which controls the maximum number of splits allowed. By using **n=1**, the method will make a single split on the first whitespace:

<img width="500" src="https://drive.google.com/uc?export=view&id=1SG2mTLap_Ym7mi1epG9U0upKN7hYL_da">

Manipulating Python lists inside a Series can be cumbersome. Instead, we can use the **expand=True** argument which will expand our series of lists into a dataframe:

<img width="500" src="https://drive.google.com/uc?export=view&id=1RKx_YuufvLXt-ei1pyu54SIQjfF7bpzY">

Lastly, we can use **DataFrame.iloc[]** to select only the first column:


<img width="500" src="https://drive.google.com/uc?export=view&id=1HjB3Ry_pnvtdj_K11bWaturCwBbyFOsz">


Let's use this technique to extract the manufacturer from the **cpu** column as well. Here are the first 10 rows of the **cpu** column, you'll see they follow a similar format to the **gpu** column:





In [0]:
laptops.cpu.head(10)

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>


1. In the example code, we have extracted the **manufacturer name** from the **gpu** column, and assigned it to a new column **gpu_manufacturer.**

  - Extract the **manufacturer name** from the **cpu** column, and assign it to a new column **cpu_manufacturer**.

In [0]:
# put your code here
laptops['cpu_manufacturer'] = laptops['cpu'].str.split(n=1, expand=True).iloc[:,0]

print(laptops['cpu_manufacturer'].unique())

['Intel' 'AMD' 'Samsung']


## 6 Extracting Values from the End of Strings


Next, let's look at the **screen** column:



In [0]:
print(laptops["screen"].unique().shape)
print(laptops["screen"].unique()[:10])

(40,)
['IPS Panel Retina Display 2560x1600' '1440x900' 'Full HD 1920x1080'
 'IPS Panel Retina Display 2880x1800' '1366x768'
 'IPS Panel Full HD 1920x1080' 'IPS Panel Retina Display 2304x1440'
 'IPS Panel Full HD / Touchscreen 1920x1080'
 'Full HD / Touchscreen 1920x1080' 'Touchscreen / Quad HD+ 3200x1800']


There are 40 unique values, so we'll just look at the first 10. Some of the values have just the resolution (eg **'1440x900'**), while others have information on the type of screen and then end with the resolution. While the resolution is mostly numeric, we're going to treat the resolution as as a substring that we want to extract. Let's start by seeing what we get when we use **Series.str.split()**:


In [0]:
laptops.loc[:9, "screen"].str.split(expand=True)

0      1          2        3          4
0       IPS  Panel     Retina  Display  2560x1600
1  1440x900   None       None     None       None
2      Full     HD  1920x1080     None       None
3       IPS  Panel     Retina  Display  2880x1800
4       IPS  Panel     Retina  Display  2560x1600
5  1366x768   None       None     None       None
6       IPS  Panel     Retina  Display  2880x1800
7  1440x900   None       None     None       None
8      Full     HD  1920x1080     None       None
9       IPS  Panel       Full       HD  1920x1080

Five of the rows have the resolution in column 4, two have it in column 2 and the rest have it in column 1. Because the value we want to extract is from the end of the string and not the beginning of the string, the **n=1** parameter is not going to help us. Luckily the **Series.str.rsplit()** method allows us to split from the end of the string instead of the front.

Let's look at a quick example to see the different between **Series.str.rsplit()** and **Series.str.split()**. We'll use a sample series called sentences:

<img width="600" src="https://drive.google.com/uc?export=view&id=1QJUgCe5cj38Xe-8Xlb8B63Vgjd1CQHOJ">

Let's use **Series.str.split()** with n=1:


<img width="600" src="https://drive.google.com/uc?export=view&id=1eipfc5ZVLJyZ5cN_Re7RQc3g5NosRz_A">


We have split the first word from the start of the string, effectively extracting a series of names. Next, let's use **Series.str.rsplit()** on the original sentences series:


<img width="600" src="https://drive.google.com/uc?export=view&id=1_ON37DLE1xegsXHFUQ9NhmBODpOO7ebr">


We have split the first word from the end of the string, effectively creating a series of colors.

Now that we understand how **Series.str.rsplit()** works, let's look at how we can use it to extract the resolutions from our screen column:

In [0]:
laptops.loc[:9, "screen"].str.rsplit(n=1, expand=True)

0          1
0  IPS Panel Retina Display  2560x1600
1                  1440x900       None
2                   Full HD  1920x1080
3  IPS Panel Retina Display  2880x1800
4  IPS Panel Retina Display  2560x1600
5                  1366x768       None
6  IPS Panel Retina Display  2880x1800
7                  1440x900       None
8                   Full HD  1920x1080
9         IPS Panel Full HD  1920x1080

This is much closer to what we're looking for - almost all of the resolution substrings are in column 1, and those that don't have the resolution in column 0 and have a null value in column 1. We can use **Series.isnull()** to assign just to the rows with the **null values**:

In [0]:
screen_res = laptops["screen"].str.rsplit(n=1, expand=True)

# giving the columns string labels makes them easier to work with
screen_res.columns = ["A", "B"]

# for rows where the value of column "B" is null, fill in the
# value found in column "A" for that row
screen_res.loc[screen_res["B"].isnull(), "B"] = screen_res["A"]

screen_res.iloc[:10]

A          B
0  IPS Panel Retina Display  2560x1600
1                  1440x900   1440x900
2                   Full HD  1920x1080
3  IPS Panel Retina Display  2880x1800
4  IPS Panel Retina Display  2560x1600
5                  1366x768   1366x768
6  IPS Panel Retina Display  2880x1800
7                  1440x900   1440x900
8                   Full HD  1920x1080
9         IPS Panel Full HD  1920x1080

We've now extracted the resolution data, and we can then assign this to a new column:



In [0]:
laptops["screen_resolution"] = screen_res["B"]
print(laptops["screen_resolution"].unique().shape)
laptops["screen_resolution"].unique()

(15,)


array(['2560x1600', '1440x900', '1920x1080', '2880x1800', '1366x768',
       '2304x1440', '3200x1800', '1920x1200', '2256x1504', '3840x2160',
       '2160x1440', '2560x1440', '1600x900', '2736x1824', '2400x1600'],
      dtype=object)

Our new **screen_resolution** column has just 15 unique values, down from the original 40. Let's use a similar technique to extract the **CPU speed** from the **cpu** column:

In [0]:
laptops["cpu"].unique()[:5]

array(['Intel Core i5 2.3GHz', 'Intel Core i5 1.8GHz',
       'Intel Core i5 7200U 2.5GHz', 'Intel Core i7 2.7GHz',
       'Intel Core i5 3.1GHz'], dtype=object)

From this data, we want to extract the processor **speed** as a number, eg from the string **'Intel Core i5 2.3GHz'** we want to extract a **float 2.3**.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- From the **cpu** column, perform the following steps to extract and convert the **processor speed**, assigning the results to a new column **cpu_speed_ghz**:
  - Use **Series.str.replace()** to remove the substring **"GHz"** each string.
  - Use **Series.str.rsplit()** and **DataFrame.iloc[]** to select the numeric characters from the end of the string.
  - Use **Series.astype()** to cast the values to the **float dtype**.

In [0]:
# put your code here
laptops['cpu_speed_ghz'] = laptops['cpu'].str.replace('GHz', '')
laptops['cpu_speed_ghz'] = laptops['cpu_speed_ghz'].str.rsplit(n=1, expand=True).iloc[:,0]
laptops['cpu_speed_ghz'] = laptops['cpu_speed_ghz'].astype(float)

ValueError: ignored

## 7 Correcting Bad Values



If your data has been scraped from a webpage, or if there was manual data entry involved at some point, you may end up with inconsistent values. Let's look at an example from our **os** column:

In [0]:
laptops.os.value_counts()

Windows      1125
No OS          66
Linux          62
Chrome OS      27
macOS          13
Mac OS          8
Android         2
Name: os, dtype: int64

We can see that there are two variations on how the Apple operating system **macOS** exists in our dataset: **Mac OS** and **macOS**. One way we could fix this is by using a boolean comparison and assignment, but instead we'll learn a new way: the [Series.map()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.map.html) method. The **Series.map()** method is ideal when we want to change multiple values in a column. Even though that's not the case here, we'll use it as an opportunity to learn how the method works.

The most common way to use **Series.map()** is with a dictionary. The keys of our dictionary are the original values in our series, and the corresponding values are what they're updated to. Let's look at a simple example using a series of misspelled fruit:



In [0]:
s = pd.Series(["pair","oranje","bananna","oranje","oranje","oranje"])

We'll create a **dictionary** called corrections and pass that dictionary as an argument to **Series.map()**:

In [0]:
corrections = {
    "pair": "pear",
    "oranje": "orange",
    "bananna": "banana"
}

s = s.map(corrections)
print(s)

0      pear
1    orange
2    banana
3    orange
4    orange
5    orange
dtype: object


We can see that each of our corrections were made across our series. One important thing to remember with **Series.map()** is that if a value from your series doesn't exist as a key in your dictionary, it will convert that value to **NaN**. Let's see what happens when we run map one more time:

In [0]:
s.map(corrections)

0    NaN
1    NaN
2    NaN
3    NaN
4    NaN
5    NaN
dtype: object

Because none of the corrected values in our series existed as keys in our dictionary, all values become **NaN**! It's very common to come across this, **especially when working in Jupyter notebook where you can easily re-run cells**.

Let's use **Series.map()** to clean the values in the **os** column.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>


1. We have created a dictionary for you to use with mapping. Note that we have included both the correct and incorrect spelling of macOS as keys, otherwise we'll end up with null values.
  - Use the **Series.map()** method with the **mapping_dict** dictionary to correct the values in the **os** column.

In [0]:
mapping_dict = {
    'Android': 'Android',
    'Chrome OS': 'Chrome OS',
    'Linux': 'Linux',
    'Mac OS': 'macOS',
    'No OS': 'No OS',
    'Windows': 'Windows',
    'macOS': 'macOS'
}

# put your code here

## 8 Dropping Missing Values



In previous missions, we've talked briefly about missing values, and how both NumPy and pandas represent these as null values. In pandas null values will be indicated by either **NaN** or **None**. Generally the first thing that we want to do is identify which values are missing.

There are two approaches we can use: the **DataFrame.info()** method and the **DataFrame.isnull()** method. The **DataFrame.info()** method will print information about the dataframe, including the number of **non-null** values in each column:

In [0]:
laptops.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 13 columns):
manufacturer                1303 non-null object
model name                  1303 non-null object
category                    1303 non-null object
screen size                 1303 non-null object
screen                      1303 non-null object
cpu                         1303 non-null object
ram                         1303 non-null object
storage                     1303 non-null object
gpu                         1303 non-null object
operating system            1303 non-null object
operating system version    1133 non-null object
weight                      1303 non-null object
price euros                 1303 non-null object
dtypes: object(13)
memory usage: 132.4+ KB


There are two downsides to this approach 
  - firstly the information is printed, so we can't easily work with it, 
  - and secondly looking at the number of non-null values can be harder to understand than looking at the number of null values. 
  
In contrast, **DataFrame.isnull()** returns a boolean dataframe with **True** and **False** indications for every value in the dataframe, and then we can use **DataFrame.sum()** to give us accounts– using a **.sum()** method on a boolean array will give us a count of the **True** values:


In [0]:
laptops.isnull().sum()

manufacturer                  0
model name                    0
category                      0
screen size                   0
screen                        0
cpu                           0
ram                           0
storage                       0
gpu                           0
operating system              0
operating system version    170
weight                        0
price euros                   0
dtype: int64

It's a lot clearer that we have only one column with null values– **os_version**, which has 170 missing values of the ~1300 total, about 13%.

We have a few options for how we can handle missing values:

- Remove any rows that have missing values.
- Remove any columns that have missing values.
- Fill the missing values with some other value.
- Leave the missing values as is.

The first two options, removing columns and/or rows with missing values is often used when preparing data for machine learning, as machine learning algorithms are unable to be trained on data that includes null values. The methods that we use to remove rows and columns with null values is the **DataFrame.dropna()** method. As a result, removing columns and rows is commonly known as **dropping**.

The **DataFrame.dropna()** method accepts an **axis** parameter, which indicates whether we want to drop along the column or index axis. Let's look at an example of this using an example dataframe:

<img width="500" src="https://drive.google.com/uc?export=view&id=1ekeSgwQGmBjbHLQwcNyWp7QyGlH9JHgE">

The default value for the axis parameter is 0, so **df.dropna()** returns an identical result to **df.dropna(axis=0)**:

<img width="500" src="https://drive.google.com/uc?export=view&id=1otmPswYfIZ0RHuHVcvYAy4hN5-Z7lqN4">


The rows with labels x and z contain null values, so those values are dropped. Let's look at what happens when we use **axis=1** to specify the column axis:


<img width="500" src="https://drive.google.com/uc?export=view&id=1ygwufYovi1TvKzhzyP5X72l96wOHyDdU">

Only the column with label C contains null values, so in this case just one column is removed.

Based off our earlier exercises to identify which columns contain null values, and how many, we know that these two techniques will remove 170 rows and 1 column, respectively. Let's practice using **DataFrame.dropna()** to remove rows and columns:


**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>



- Use **DataFrame.dropna()** to remove any rows from the **laptops** dataframe that have null values, assigning the result to **laptops_no_null_rows**.
- Use **DataFrame.dropna()** to remove any columns from the **laptops** dataframe that have null values, assigning the result to **laptops_no_null_cols**.


In [0]:
# put your code here

## 9 Filling Missing Values




In the previous screen, we learned there are various ways we can choose to deal with missing values:

- Remove any rows that have missing values.
- Remove any columns that have missing values.
- Fill the missing values with some other value.
- Leave the missing values as is.

While choosing to drop either the **rows** or **columns** is the easiest approach to dealing with missing values, it may not always be the best approach. If, for example, one particular manufacturer's laptops have a greater percentage of missing values for the **os_version** column, we might have removed a disproportionate amount of that manufacturer's laptops, which would affect our analysis.

Because of this, it's worthwhile exploring the missing values before you make your decision. One method is to explore all of the values in the column. We can use **Series.value_counts()** for this, but we'll use a new parameter we haven't previously used, **dropna=False**. By default, **Series.value_counts()** won't include null values in its output. This parameter allows us to explicitly indicate we want to see the null values:



In [0]:
laptops["os_version"].value_counts(dropna=False)

We can see that the majority of values in the column are **10**, with the missing values the next most commons, and then about 5% of values being one of three others.

We can also explore values of the other columns in the rows with null values. In this case, the **os_version** column is closely related to the os column, so we'll look at those values:


In [0]:
os_with_null_v = laptops.loc[laptops["os_version"].isnull(),"os"]
os_with_null_v.value_counts()

Immediately we can observe a few things:

- Most of the missing values are actually when the laptop doesn't include any OS. This is an important distinction, because it's not so much that we don't know what the value is, as that there can't be a value.
- 13 of the laptops that come with macOS do not specify the version. Leaning on our knowledge of MacOS, we might know that the full name of macOS used to be Mac OS X, and so we might to fill these values to be more consistent.

In both of these cases, we can fill the missing values to make our data more correct. For the rest of the values, it's probably best to leave them as missing so we don't remove important values.

First, let's explore those mac columns a bit more to make sure our intuition was correct

In [0]:
mac_os_versions = laptops.loc[laptops["os"] == "macOS", "os_version"]
mac_os_versions.value_counts(dropna=False)

In [0]:
laptops.loc[laptops["os"] == "macOS", "os_version"] = "X"

For our other case, let's insert a **No OS** value into the **os_version** column for any laptop with a **No OS** value in the **os** column.


**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>


- Use a boolean array to identify rows that have the value **No OS** for the **os** column, and then use assignment to assign the value **Version Unknown** to the **os_version** column for those rows.
- Use the syntax below to create **value_counts_after** variable:

```python
value_counts_after = laptops.loc[laptops["os_version"].isnull(), "os"].value_counts()
```


In [0]:
# put your code here

## 10 Challenge: Extracting Storage Information



Now it's time for a challenge to bring together a lot of the concepts we've learned so far! In this challenge we're going to clean the storage column, which contains information about the disks within the laptops. Let's look at a sample of the data in that column:


In [0]:
laptops.loc[76:81, "storage"]

76                 2TB HDD
77    128GB SSD +  1TB HDD
78                 1TB HDD
79    128GB SSD +  1TB HDD
80               256GB SSD
81               512GB SSD
Name: storage, dtype: object

From this sample, we can observe:

- Some laptops have two disks and some just have one.
- Each disk has a capacity (eg 128GB) and a type (eg SSD).
- Capacities are expressed in either gigabytes (GB) or terabytes (TB).

Our aim is to extract this data into several new columns. Our final data will looks like this:


| _  | storage_1_capacity_gb | storage_1_type | storage_2_capacity_gb | storage_2_type |
|----|-----------------------|----------------|-----------------------|----------------|
| 76 | 2000.0                | HDD            | NaN                   | None           |
| 77 | 128.0                 | SSD            | 1000.0                | HDD            |
| 78 | 1000.0                | HDD            | NaN                   | None           |
| 79 | 128.0                 | SSD            | 1000.0                | HDD            |
| 80 | 256.0                 | SSD            | NaN                   | None           |
| 81 | 512.0                 | SSD            | NaN                   |  None              |


- The **storage_1_capacity_gb** and **storage_2_capacity_gb** columns are both **float** type, and both store the capacity with uniform units - gigabytes, where 1TB is 1000GB.
- If there is only one disk in the laptop, both **storage_2_capacity_gb** and **storage_2_type** are null.


Your task is to create the four new columns as seen above, and remove the existing storage column. To remove a column, you can use the **DataFrame.drop()** method:

```python
laptops = laptops.drop('storage', axis=1)
```

If you create any temporary columns along the way in order to complete these challenge, **DataFrame.drop()** accepts a list of labels so you can remove them all in one line.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

1. Clean the **storage** column, creating four new columns:
  - **storage_1_capacity_gb**, with **float** dtype.
  - **storage_1_type**.
  - **storage_2_capacity_gb**, with **float** dtype. If there is only one drive, this column should be null.
  - **storage_2_type**. If there is only one drive, this column should be null.
  - If needed, don't forget to strip the columns of any extra whitespace.
2. Drop the **original storage** column and any temporary columns you made while completing the exercise.


In [0]:
# tip

# replace 'TB' with 000 and rm 'GB'
laptops["storage"] = laptops["storage"].str.replace('GB','').str.replace('TB','000')

# split out into two columns for storage
laptops[["storage_1", "storage_2"]] = laptops["storage"].str.split("+", expand=True)

# put your code here

## 11 Reordering Columns and Exporting Cleaned Data



We have finished cleaning our data. Our last step is to reorder the columns, and then save out our cleaned data to a CSV file, so that we can re-read it again as a clean version.

To reorder columns, we start by using the **DataFrame.columns** attribute to produce a list of the current columns:

```python
print(laptops.columns)

['manufacturer', 'model_name', 'category', 'screen_size_inches',
 'screen', 'cpu', 'ram_gb', 'gpu', 'os', 'os_version',
 'weight_kg', 'price_euros', 'cpu_manufacturer',
 'screen_resolution', 'cpu_speed', 'storage_1_capacity_gb',
 'storage_1_type', 'storage_2_capacity_gb', 'storage_2_type']
```

All of our new columns have been added to the end of the axis– our aim is to group similar columns together. We can manually rearrange this list, and then assign the rearranged list back to laptops.column.

To save out our CSV file, we use the **DataFrame.to_csv()** method. The syntax you'll use most of the time when saving out a CSV is:

```python
df.to_csv('filename.csv', index=False)
```

By default, pandas will save the index labels as a column in the CSV file. Our dataset has integer labels which don't contain any data, so we don't need to save the index out.

Let's re-order the columns, save out the cleaned CSV, and then re-read it into a new dataframe to compare the dtypes before and after.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Using the **cols** list we have created for laptops, reorder the columns in the laptops dataframe.
- Use the **DataFrame.to_csv()** method to save the laptops dataframe to a CSV file **laptops_cleaned.csv** without index labels.
- Read **laptops_cleaned.csv** into a new dataframe, **laptops_cleaned**.
- Use the **DataFrame.dtypes** attribute to assign the dtypes from **laptops_cleaned** to **laptops_cleaned_dtypes**.


In [0]:
laptops_dtypes = laptops.dtypes
cols = ['manufacturer', 'model_name', 'category', 'screen_size_inches',
        'screen', 'cpu', 'cpu_manufacturer',  'cpu_speed', 'ram_gb',
        'storage_1_type', 'storage_1_capacity_gb', 'storage_2_type',
        'storage_2_capacity_gb', 'gpu', 'gpu_manufacturer', 'os',
        'os_version', 'weight_kg', 'price_euros']

# put your code here

## 12 Next steps

When we re-read in our dataframe, we got identical dtypes! It's often a good idea to save out a CSV when you finish cleaning to save you time should you wish to do analysis later.

Our dataset is now primed for analysis. Here are some questions you might like to answer in your own time by analyzing the cleaned data:

- Are laptops made by Apple more expensive than those by other manufacturers?
- What is the best value laptop with a screen size of 15" or more?
- Which laptop has the most storage space? 